packages/cli/command_help.pony

```diff
@@ -0,0 +1,198 @@
+    match _parent
+    | let p: CommandHelp => p.fullname() + " " + _spec.name()
+    else
+     _spec.name()
+    end
+
+  fun box string(): String => fullname()
+
+  fun box help_string(): String =>
+    let w: StringWriter ref = StringWriter
+    print_help(w)
+    w.string()
+
+  fun box print_help(w: Writer) =>
+    print_usage(w)
+    if _spec.descr().size() > 0 then
+      w.write("\n")
+      w.write(_spec.descr() + "\n")
+    end
+
+    let options = all_options()
+    if options.size() > 0 then
+      w.write("\nOptions:\n")
+      print_options(w, options)
+    end
+    if _spec.commands().size() > 0 then
+      w.write("\nCommands:\n")
+      print_commands(w)
+    end
+    let args = _spec.args()
+    if args.size() > 0 then
+      w.write("\nArgs:\n")
+      print_args(w, args)
+    end
+
+  fun box print_usage(w: Writer) =>
+    w.write("usage: " + fullname())
+    if any_options() then
+      w.write(" [<options>]")
+    end
+    if _spec.commands().size() > 0 then
+      w.write(" <command>")
+    end
+    if _spec.args().size() > 0 then
+      for a in _spec.args().values() do
+        w.write(" " + a.help_string())
+      end
+    else
+      w.write(" [<args> ...]")
+    end
+    w.write("\n")
+
+  fun box print_options(w: Writer, options: Array[OptionSpec box] box) =>
+    let cols = Array[(String,String)]()
+    for o in options.values() do
+      cols.push(("  " + o.help_string(), o.descr()))
+    end
+    _Columns.print(w, cols)
+
+  fun box print_commands(w: Writer) =>
+    let cols = Array[(String,String)]()
+    _list_commands(_spec, cols, 1)
+    _Columns.print(w, cols)
+
+  fun box _list_commands(
+    cs: CommandSpec box,
+    cols: Array[(String,String)],
+    level: USize)
+  =>
+    for c in cs.commands().values() do
+      cols.push((_Columns.indent(level*2) + c.help_string(), c.descr()))
+      _list_commands(c, cols, level + 1)
+    end
+
+  fun box print_args(w: Writer, args: Array[ArgSpec] box) =>
+    let cols = Array[(String,String)]()
+    for a in args.values() do
+      cols.push(("  " + a.help_string(), a.descr()))
+    end
+    _Columns.print(w, cols)
+
+  fun box any_options(): Bool =>
+    if _spec.options().size() > 0 then
+      true
+    else
+      match _parent
+      | let p: CommandHelp => p.any_options()
+      else
+        false
+      end
+    end
+
+  fun box all_options(): Array[OptionSpec box] =>
+    let options = Array[OptionSpec box]()
+    _all_options(options)
+    options
+
+  fun box _all_options(options: Array[OptionSpec box]) =>
+    match _parent
+    | let p: CommandHelp => p._all_options(options)
+    end
+    for o in _spec.options().values() do
+      options.push(o)
+    end
+
+interface Writer
+  """
+  This interface and two classes allow the help output to go to a general
+  Writer that can be implemented for string creation or to an OutStream. Or
+  other targets. This could be polished up and moved into a more central
+  package.
+  """
```

Comment:
I still don't really think that `Writer`, `StringWriter`, and `OutWriter` belong as public types provided by this package. As you recognize in the docstring, they're not intrinsically related to CLI concerns.

Also, for types/concepts as fundamental as these I'd *really* like to see some robust RFC discussion about the best approach to provide a consistent API layer for this concern. It really is a big enough discussion that it deserves to be separate from this one, and I can already think of some concerns I have about the design as it's presented here (for example, it doesn't account for `writev` needs, and the many-`append` approach of `StringWriter` is not particularly optimized).

Instead of slipping them in here and migrating them to a different package later, I'd prefer to see them removed from here and added later, once the design has been worked out.

For the time being, I'd suggest refactoring to use the `Writer` type from the `buffered` package, which already exists and serves a similar purpose to that of your `StringWriter`.  For the public methods that accept your `Writer` type as an argument, I suggest something like this:

```pony
interface _Writev
  be writev(data: ByteSeqIter)
```
```pony
class box CommandHelp
  // ...

  fun help_string(): String =>
    let w: buffered.Writer = buffered.Writer
    _write_help(w)
    let size = w.size()
    let string = recover trn String(size) end
    for bytes in s.done().values() do string.append(bytes) end
    consume string

  fun print_help(os: OutStream) =>
    let w: buffered.Writer = buffered.Writer
    _write_help(w)
    os.writev(w.done())

  fun _write_help(w: buffered.Writer) =>
    print_usage(w)
    if _spec.descr().size() > 0 then
      w.write("\n")
      w.write(_spec.descr() + "\n")
    end

    let options = all_options()
    if options.size() > 0 then
      w.write("\nOptions:\n")
      print_options(w, options)
    end
    if _spec.commands().size() > 0 then
      w.write("\nCommands:\n")
      print_commands(w)
    end
    let args = _spec.args()
    if args.size() > 0 then
      w.write("\nArgs:\n")
      print_args(w, args)
    end

  // ...
```